src/components/shared/Swiper.tsx

```diff
@@ -0,0 +1,329 @@
+  /**
+   * Initialize the state
+   */
+  initState(props: SwiperProps): SwiperState {
+    // Get the total number of slides passed as children
+    const total = props.children ? props.children.length || 1 : 0,
+      // Current index
+      index =
+        total > 1 ? Math.min(props.index ? props.index : 0, total - 1) : 0,
+      // Current offset
+      offset = width * index;
+
+    const state = {
+      total,
+      index,
+      offset,
+      width,
+      height,
+      isScrolling: false
+    };
+
+    return state;
+  }
+
+  /**
+   * Scroll begin handler
+   * @param {object} e native event
+   */
+  onScrollBegin: ScrollViewProps["onScrollBeginDrag"] = (
+    event?: NativeSyntheticEvent<NativeScrollEvent>
+  ) => {
+    if (!event) {
+      // TODO: what does this case even mean? read react-native docs
+      return;
+    }
+    // Update internal isScrolling state
+    this.setState({ isScrolling: true });
+  };
+
+  /**
+   * Scroll end handler
+   */
+  onScrollEnd: ScrollViewProps["onMomentumScrollEnd"] = event => {
+    if (!event) {
+      // TODO: what does this case even mean? read react-native docs
+      return;
+    }
+    // Update internal isScrolling state
+    this.state.isScrolling = false;
+
+    // Update index
+    this.updateIndex(
+      event.nativeEvent.contentOffset
+        ? event.nativeEvent.contentOffset.x
+        : this.state.offset
+      // TODO: When scrolled with .scrollTo() on Android there is no contentOffset
+      // event.nativeEvent.position * this.state.width
+    );
+  };
+
+  /*
+     * Drag end handler
+     */
+  onScrollEndDrag: ScrollViewProps["onScrollEndDrag"] = event => {
+    if (!event) {
+      // TODO: what does this case even mean? read react-native docs
+      return;
+    }
+    const {
+        contentOffset: { x: newOffset }
+      } = event.nativeEvent,
+      { children } = this.props,
+      { index } = this.state,
+      { offset } = this.state;
+
+    // Update internal isScrolling state
+    // if swiped right on the last slide
+    // or left on the first one
+    if (
+      offset === newOffset &&
+      (index === 0 || index === children.length - 1)
+    ) {
+      this.setState({
+        isScrolling: false
+      });
+    }
+  };
+
+  /**
+   * Update index after scroll
+   * @param {object} offset content offset
+   */
+  updateIndex = (offset: number) => {
+    const state = this.state,
+      diff = offset - this.state.offset,
+      step = state.width;
+    let index = state.index;
+
+    // Do nothing if offset didn't change
+    if (!diff) {
+      return;
+    }
+
+    // Make sure index is always an integer
+    index = index + Math.round(diff / step);
+
+    // Update internal offset
+    this.setState({
+      offset: offset
+    });
+    // Update index in the state
+    this.setState({
+      index
+    });
+  };
+
+  /**
+   * Swipe one slide forward
+   */
+  swipe = () => {
+    // Ignore if already scrolling or if there is less than 2 slides
```

Comment:
This comment is likely to become out of date since it just describes the code, I think no need for it.